sdk/database/newdbplugin/middleware.go

```diff
@@ -0,0 +1,273 @@
+
+		if err != nil {
+			metrics.IncrCounter([]string{"database", "CreateUser", "error"}, 1)
+			metrics.IncrCounter([]string{"database", mw.typeStr, "CreateUser", "error"}, 1)
+		}
+	}(time.Now())
+
+	metrics.IncrCounter([]string{"database", "CreateUser"}, 1)
+	metrics.IncrCounter([]string{"database", mw.typeStr, "CreateUser"}, 1)
+	return mw.next.NewUser(ctx, req)
+}
+
+func (mw databaseMetricsMiddleware) UpdateUser(ctx context.Context, req UpdateUserRequest) (resp UpdateUserResponse, err error) {
+	defer func(now time.Time) {
+		metrics.MeasureSince([]string{"database", "UpdateUser"}, now)
+		metrics.MeasureSince([]string{"database", mw.typeStr, "UpdateUser"}, now)
+
+		if err != nil {
+			metrics.IncrCounter([]string{"database", "UpdateUser", "error"}, 1)
+			metrics.IncrCounter([]string{"database", mw.typeStr, "UpdateUser", "error"}, 1)
+		}
+	}(time.Now())
+
+	metrics.IncrCounter([]string{"database", "UpdateUser"}, 1)
+	metrics.IncrCounter([]string{"database", mw.typeStr, "UpdateUser"}, 1)
+	return mw.next.UpdateUser(ctx, req)
+}
+
+func (mw databaseMetricsMiddleware) DeleteUser(ctx context.Context, req DeleteUserRequest) (resp DeleteUserResponse, err error) {
+	defer func(now time.Time) {
+		metrics.MeasureSince([]string{"database", "DeleteUser"}, now)
+		metrics.MeasureSince([]string{"database", mw.typeStr, "DeleteUser"}, now)
+
+		if err != nil {
+			metrics.IncrCounter([]string{"database", "DeleteUser", "error"}, 1)
+			metrics.IncrCounter([]string{"database", mw.typeStr, "DeleteUser", "error"}, 1)
+		}
+	}(time.Now())
+
+	metrics.IncrCounter([]string{"database", "DeleteUser"}, 1)
+	metrics.IncrCounter([]string{"database", mw.typeStr, "DeleteUser"}, 1)
+	return mw.next.DeleteUser(ctx, req)
+}
+
+func (mw databaseMetricsMiddleware) Type() (string, error) {
+	return mw.next.Type()
+}
+
+func (mw databaseMetricsMiddleware) Close() (err error) {
+	defer func(now time.Time) {
+		metrics.MeasureSince([]string{"database", "Close"}, now)
+		metrics.MeasureSince([]string{"database", mw.typeStr, "Close"}, now)
+
+		if err != nil {
+			metrics.IncrCounter([]string{"database", "Close", "error"}, 1)
+			metrics.IncrCounter([]string{"database", mw.typeStr, "Close", "error"}, 1)
+		}
+	}(time.Now())
+
+	metrics.IncrCounter([]string{"database", "Close"}, 1)
+	metrics.IncrCounter([]string{"database", mw.typeStr, "Close"}, 1)
+	return mw.next.Close()
+}
+
+// ///////////////////////////////////////////////////
+// Error Sanitizer Middleware Domain
+// ///////////////////////////////////////////////////
+
+var _ Database = DatabaseErrorSanitizerMiddleware{}
+
+// DatabaseErrorSanitizerMiddleware wraps an implementation of Databases and
+// sanitizes returned error messages
+type DatabaseErrorSanitizerMiddleware struct {
+	next      Database
+	secretsFn secretsFn
+}
+
+type secretsFn func() map[string]string
+
+func NewDatabaseErrorSanitizerMiddleware(next Database, secrets secretsFn) DatabaseErrorSanitizerMiddleware {
```

Comment:
I think I know why we're returning a concrete object here (not really modifying the object's fields after instantiation) and why we're not dealing with pointer receivers for its methods, but can you elaborate to confirm my assertions?

Since this is a middleware and likely to be passed around and reference as an interface, I am not sure if this can be problematic upstream at some point if the caller expects a pointer value.

Reply:
You nailed it: the middleware isn't changing after it is initially created. Passing it around under an interface type won't change any behaviors here. All interface types eg `var db Database` are implicitly pointers so you can do things like nil checks, however the underlying object such as `DatabaseErrorSanitizerMiddleware` can be a concrete, non-pointer type without issue. Since the middleware type is never manipulating its own variables in any of its functions this won't cause any problems.

Reply:
Are there any cases where you'd need to type assert on the interface to a typed pointer value followed by a nil check?

Reply:
None that I've found. All of the usages of this use this as the `Database` interface, not as the underlying middleware type.